In [8]:
import pandas as pd

df = pd.read_csv("data/customer_data_cleaned.csv")
print("Rows loaded:", len(df))
print("Columns:", df.columns.tolist())

Rows loaded: 500
Columns: ['customer_id', 'signup_date', 'age', 'gender', 'city', 'segment', 'income', 'purchase_amount', 'satisfaction_score', 'age_was_missing', 'income_was_missing', 'age_band', 'signup_year', 'signup_month', 'income_band']


## Core Metrics — Definitions

1. **Total Customers** — Count of unique, de-duplicated customer records in the cleaned dataset.
2. **Average Income** — Mean of the `income` column, calculated only on non-missing, non-placeholder values (already handled in Task 4 cleaning).
3. **Average Purchase Amount** — Mean of `purchase_amount` across all customers.
4. **Average Satisfaction Score** — Mean of `satisfaction_score` across all customers.
5. **Customer Count by Segment** — Count of customers grouped by `segment` (Retail/Corporate/SME).

## Aggregation Grain

- Customer-level metrics (count, averages): one row per unique customer.
- Segment breakdown: one row per segment (3 total).

In [9]:

df['segment'] = df['segment'].replace({'Sme': 'SME'})

total_customers_rows = len(df)
total_customers_unique_id = df['customer_id'].nunique()

avg_income = df['income'].mean()
avg_purchase = df['purchase_amount'].mean()
avg_satisfaction = df['satisfaction_score'].mean()
segment_counts = df['segment'].value_counts()

print("Total Customers (row-based, recommended):", total_customers_rows)
print("Total Customers (unique ID, known to undercount):", total_customers_unique_id)
print("Average Income: ₹{:,.0f}".format(avg_income))
print("Average Purchase Amount: ₹{:,.0f}".format(avg_purchase))
print("Average Satisfaction Score:", round(avg_satisfaction, 2))
print("\nCustomers by Segment:")
print(segment_counts)

Total Customers (row-based, recommended): 500
Total Customers (unique ID, known to undercount): 492
Average Income: ₹56,586
Average Purchase Amount: ₹3,027
Average Satisfaction Score: 2.89

Customers by Segment:
segment
Retail       233
Corporate    134
SME          133
Name: count, dtype: int64


In [10]:
# Manual cross-check: does segment count add up to total?
print("Segment counts sum:", segment_counts.sum())
print("Matches total customers?", segment_counts.sum() == total_customers_rows)

# Manual cross-check: recompute average income a different way
manual_avg_income = df['income'].sum() / len(df)
print("Manual avg income check:", round(manual_avg_income, 2))
print("Matches pandas mean?", round(manual_avg_income, 2) == round(avg_income, 2))

Segment counts sum: 500
Matches total customers? True
Manual avg income check: 56585.75
Matches pandas mean? True


## Metric Dictionary — Baseline Report

| Metric | Definition | Value | Source/Grain |
|---|---|---|---|
| Total Customers | Row count in `customer_data_cleaned.csv` (post-dedup) | 500 | One row = one customer record |
| Average Income | Mean of `income` column, nulls/placeholder already excluded in Task 4 cleaning | ₹56,586 | Customer-level |
| Average Purchase Amount | Mean of `purchase_amount` column | ₹3,027 | Customer-level |
| Average Satisfaction Score | Mean of `satisfaction_score` column (1-5 scale) | 2.89 | Customer-level |
| Customers by Segment | Count of customers grouped by `segment` | Retail: 233, Corporate: 134, SME: 133 | Segment-level (one row per segment) |

**Known limitation:** `customer_id` cannot currently be used as a reliable unique-customer count (`nunique()` returns 492, undercounting due to ID-reuse issues identified in Task 2 — two different-looking customer records sharing the same ID). Row count (500) is used instead as the trusted "Total Customers" figure until the ID-reuse issue is investigated and resolved.

**Validation performed:** Segment counts sum to 500, matching total row count. Manually recomputed average income via `sum()/count()` matches pandas' `.mean()` exactly.

## Baseline Report — Published

| Metric | Value |
|---|---|
| Total Customers | 500 |
| Average Income | ₹56,586 |
| Average Purchase Amount | ₹3,027 |
| Average Satisfaction Score | 2.89 / 5 |
| Retail Customers | 233 |
| Corporate Customers | 134 |
| SME Customers | 133 |